persist energyUsage and fix some issue.

liizfq opened the pull request:
## Description

Please carefully read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md) before making any pull requests.
And, **Do not make a pull request to merge into stable unless it is a hotfix. Use the development branch instead.**
## Issues fixed by this PR

<!--- Put the links of issues that may be fixed by this PR here (if any). -->
## Type of changes

<!--- Put an `x` in all the boxes that apply your changes. -->

- [ ] Bug fix
- [ ] New feature 
- [x] Enhancement
- [ ] Documentation

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] My pull request is unique and no other pull requests have been opened for these changes
- [ ] I have read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md)
- [ ] I am responsible for any copyright issues with my code if it occurs in the future.

liizfq committed:
add new command (unlimitenergy):toggle energyusage for each player while  energyusage is ture in config.json

liizfq committed:
Solve the problem of layout and naming errors

liizfq committed:
make currentActiveTeam's Avatar full-energy while turn on the ule.

liizfq committed:
Resolve language document errors

liizfq committed:
add config_error message while player try to execute UnlimitEnergyCommand in GAME_OPTIONS.energyUsage == false

liizfq committed:
Merge branch 'Grasscutters:development' into development

liizfq committed:
Merge branch 'Grasscutters:development' into development

liizfq committed:
Merge branch 'Grasscutters:development' into development

liizfq committed:
Merge branch 'Grasscutters:development' into development

liizfq committed:
Merge branch 'Grasscutters:development' into development

liizfq committed:
1.remove energyUsage from config.json and EnergyManager.Make ule command is useful in all energyUsage status.
2.add energyUsage into player and mongodb.
3.solve char not full energy from backstage into team while ule on.
4.favonius products while balls.

GanyusLeftHorn commented on src/main/java/emu/grasscutter/game/avatar/Avatar.java:
```diff
@@ -359,30 +360,34 @@ public void setCurrentHp(float currentHp) {
-		if (GAME_OPTIONS.energyUsage) {
-			this.setFightProperty(curEnergyProp, currentEnergy);
-			this.currentEnergy = currentEnergy;
+		this.setFightProperty(curEnergyProp, currentEnergy);
+		this.currentEnergy = currentEnergy;
+		if (this.getPlayer().getEnergyUsage()) {
```
This sets the energy irrespective of whether energy usage is turned on or off. That seems like a bug?

GanyusLeftHorn commented on src/main/java/emu/grasscutter/game/entity/EntityAvatar.java:
```diff
@@ -175,12 +175,28 @@ public void addEnergy(float amount, PropChangeReason reason, boolean isFlat) {
+	}
+
+	private void setCurrentEnergy() {
+		float curEnergy;
+		FightProperty curEnergyProp = this.getAvatar().getSkillDepot().getElementType().getCurEnergyProp();
+		FightProperty maxEnergyProp = this.getAvatar().getSkillDepot().getElementType().getMaxEnergyProp();
+		PropChangeReason reason=PropChangeReason.PROP_CHANGE_REASON_ABILITY;
+		if(this.getPlayer().getEnergyUsage()){
+			curEnergy=this.avatar.getCurrentEnergy();
+		}else{
+			curEnergy= this.getFightProperty(maxEnergyProp);
 		}
-	} 
+		setCurrentEnergy(curEnergyProp,curEnergy,reason);
+	}
+
+	private void setCurrentEnergy(FightProperty curEnergyProp,float newEnergy,PropChangeReason reason) {
+		this.avatar.setCurrentEnergy(curEnergyProp, newEnergy);
+		this.getScene().broadcastPacket(new PacketAvatarFightPropUpdateNotify(this.getAvatar(), curEnergyProp));
+		this.getScene().broadcastPacket(new PacketEntityFightPropChangeReasonNotify(this, curEnergyProp, newEnergy, reason));
+	}
```
It is confusing that there is now a `setCurrentEnergy` for both `Avatar` and `EntityAvatar`. Please unify them and put them into either one of those classes, probably in `EntityAvatar`.

GanyusLeftHorn commented on src/main/java/emu/grasscutter/game/managers/energy/EnergyManager.java:
```diff
@@ -152,12 +187,19 @@ public void handleGenerateElemBall(AbilityInvokeEntry invoke) throws InvalidProt
         // Generate 2 particles by default.
-        int amount = 2;
+        int amount = particlesDefault;
+        if(invoke.getHead().getLocalId()==98307){
+            amount = 3;
+            for (int i = 0; i < amount; i++) {
+                generateElemBall(itemId, new Position(action.getPos()), 1);
+            }
+            return;
+        }
```
What is that? A fix for Favonius? Please put a discriptive comment if you handle a specific special case.

GanyusLeftHorn commented on src/main/java/emu/grasscutter/utils/ConfigContainer.java:
```diff
@@ -176,7 +176,7 @@ public static class GameOptions {
         public boolean enableShopItems = true;
         public boolean staminaUsage = true;
-        public boolean energyUsage = false;
+//        public boolean energyUsage = false;
```
I don't get what the point or benefit of this change is supposed to be. On the contrary, I can only see this as a detriment. What does taking away server manager's ability to set a default for whether they want to use such a - highly experimental, by the way - feature improve?

Giving the ability to override this config with a command is fine. Persisting that is fine too (though you would have to devise a way to differentiate between nothing being set and the feature being turned off). But taking away the configurability is not fine.

GanyusLeftHorn commented on src/main/java/emu/grasscutter/game/avatar/Avatar.java:
```diff
@@ -359,30 +360,34 @@ public void setCurrentHp(float currentHp) {
+	public float getCurrentEnergy() {
+		return currentEnergy;
 	}
+
+//	public void setCurrentEnergy() {
+//		if (this.getPlayer().getEnergyUsage()) {
+//			this.setCurrentEnergy(this.currentEnergy);
+//		}
+//	}
```
Leftover.

GanyusLeftHorn commented on src/main/java/emu/grasscutter/game/entity/EntityAvatar.java:
```diff
@@ -175,12 +175,28 @@ public void addEnergy(float amount, PropChangeReason reason, boolean isFlat) {
+		FightProperty curEnergyProp = this.getAvatar().getSkillDepot().getElementType().getCurEnergyProp();
+		FightProperty maxEnergyProp = this.getAvatar().getSkillDepot().getElementType().getMaxEnergyProp();
+		PropChangeReason reason=PropChangeReason.PROP_CHANGE_REASON_ABILITY;
+		if(this.getPlayer().getEnergyUsage()){
+			curEnergy=this.avatar.getCurrentEnergy();
+		}else{
+			curEnergy= this.getFightProperty(maxEnergyProp);
 		}
-	} 
+		setCurrentEnergy(curEnergyProp,curEnergy,reason);
```
This violates the protocol. On avatar creation, when this is called by the constructor, the live server sends neither an `AvatarFightPropUpdateNotify` nor an `EntityFightPropChangeReasonNotify`.

liizfq committed:
1.Change ule commande  execution method.
2.split setCurrentEnergy(float currentEnergy) into  setCurrentEnergy and setMaxEnergy.
3.Withdrew some of my unreasonable changes

GanyusLeftHorn commented on src/main/java/emu/grasscutter/game/entity/EntityAvatar.java:
```diff
@@ -175,12 +182,13 @@ public void addEnergy(float amount, PropChangeReason reason, boolean isFlat) {
-			this.getScene().broadcastPacket(new PacketAvatarFightPropUpdateNotify(this.getAvatar(), curEnergyProp));
-			this.getScene().broadcastPacket(new PacketEntityFightPropChangeReasonNotify(this, curEnergyProp, newEnergy, reason));
+			this.avatar.setCurrentEnergy(newEnergy);
+			this.getScene().broadcastPacket(new PacketEntityFightPropUpdateNotify(this, curEnergyProp));
```
Adding energy should send `AvatarFightPropUpdateNotify`.

liizfq replied:
ok,but why send AvatarFightPropUpdateNotify before setCurrentEnergy?

GanyusLeftHorn replied:
What do you mean?

GanyusLeftHorn replied:
What do you mean?

liizfq replied:
i guess that this.avatar.setCurrentEnergy(newEnergy)  executes before AvatarFightPropUpdateNotify.
AvatarFightPropUpdateNotify should notify the newEnergy.

GanyusLeftHorn replied:
> i guess that this.avatar.setCurrentEnergy(newEnergy) executes before AvatarFightPropUpdateNotify.

Correct.

> AvatarFightPropUpdateNotify should notify the newEnergy.

Also correct.

I know it's slightly confusing, but there is a difference in how the official server handles communicating energy changes between adding and clearing energy. For adding energy, it sends `AvatarFightPropUpdateNotify`, and for clearing it sends `EntityFightPropUpdateNotify `

liizfq committed:
send AvatarFightPropUpdateNotify for addEnergy

liizfq committed:
Merge branch 'development' into development

KingRainbow44 committed:
Formatting changes (check #1303 for details) (#1313)

Akka0 committed:
fix gadget infinity create

GanyusLeftHorn committed:
Proper Dungeon Rewards (#1286)

* (WIP) Start with dungeon rewards.

* Add multiplayer multiplier and item probabilities.

* Add more dungeons

liizfq committed:
Merge branch 'development' into development

liizfq committed:
Update UnlimitEnergyCommand.java

liizfq committed:
Merge branch 'development' into development

liizfq committed:
resolve conflict，all of files is copy from up to date.

1-lavender-1 commented:
Why is this PR closed？I consider these new features are quite useful

liizfq commented:
> Why is this PR closed？I consider these new features are quite useful

i tried to fix conflict,but i failed.So i plan fork the lasted repository and add the new features.thanks for your reply.
